### PR TITLE
Block sorting

### PR DESCRIPTION
**Pull request by jeking04**
This PR integrates the perception pipeline created in the prpy PR by the same name. It also includes the TSRs and actions for block sorting.

### CODE REVIEW

File: src/herbpy/action/blocks.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+from prpy.planning.base import PlanningError
+from prpy.util import ComputeEnabledAABB
+
+logger = logging.getLogger('herbpy')
+
+class NoTSRException(Exception):
+    pass
+
+def _GrabBlock(robot, blocks, table, manip=None, preshape=[1.7, 1.7, 0.2, 2.45],
+              **kw_args):
```
Comment (mkoval) [inline src/herbpy/action/blocks.py]: `preshape` should default to `None` and get set to the default value inside the function.

File: src/herbpy/action/blocks.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    """
+    @param robot The robot performing the grasp
+    @param block The block to grab
+    @param table The table, or object, the block is resting on
+    @param manip The manipulator to use to grab the block
+      (if None active manipulator is used)
```
Comment (mkoval) [inline src/herbpy/action/blocks.py]: Missing `__docstring__` for `preshape`.

File: src/herbpy/action/blocks.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    """
+    from prpy.rave import AllDisabled, Disabled
+    from prpy.viz import RenderTSRList, RenderVector
+
+    env = robot.GetEnv()
+    block = None
+
+    if manip is None:
+        manip = robot.GetActiveManipulator()
```
Comment (mkoval) [inline src/herbpy/action/blocks.py]: Missing lock.

File: src/herbpy/action/blocks.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    
+    # First move the hand to the right preshape
+    manip.hand.MoveHand(*preshape)
+
+    block_tsr_list = []
+    for b in blocks:
+        # Get a TSR to move near the block.
+        tsr_list = robot.tsrlibrary(b, 'grasp', manip=manip)
+        block_tsr_list += tsr_list
```
Comment (mkoval) [inline src/herbpy/action/blocks.py]: Missing lock.

File: src/herbpy/action/blocks.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            hand_pose = manip.GetEndEffectorTransform()
+            block_pose = numpy.dot(hand_pose, block_relative)
+            block.SetTransform(block_pose)
+            manip.GetRobot().Grab(block)
+
+    except PlanningError as e:
+        logger.error('Failed to complete block grasp')
+        raise
+    finally:
+        return block
```
Comment (mkoval) [inline src/herbpy/action/blocks.py]: I don't think it makes sense to put this `return` in the `finally` block. I would remove the `finally` block entirely and put it in the function scope.

File: src/herbpy/action/blocks.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    @param manip The manipulator used to perform the placement
+    If none, the active manipulator is used.
+    """
+    env = robot.GetEnv()
+
+    # Get a tsr for this position
+    object_place_list = robot.tsrlibrary(on_obj, 'point_on', manip=manip)
+    place_tsr_list = robot.tsrlibrary(block, 'place_on', 
+                                      pose_tsr_chain=object_place_list[0], 
+                                      manip=manip)
```
Comment (mkoval) [inline src/herbpy/action/blocks.py]: Missing lock.

File: src/herbpy/herbrobot.py [hunk 7]
```diff
@@ -178,11 +178,26 @@ def __init__(self, left_arm_sim, right_arm_sim, right_ft_sim,
...
+                                            marker_data_path=marker_data_path,
+                                            kinbody_path=kinbody_path,
+                                            detection_frame='head/kinect2_rgb_optical_frame',
+                                            destination_frame='map')
         if not self.talker_simulated:
             # Initialize herbpy ROS Node
             import rospy
             if not rospy.core.is_initialized():
-                rospy.init_node('herbpy', anonymous=True)
+#                rospy.init_node('herbpy', anonymous=True)
```
Comment (mkoval) [inline src/herbpy/herbrobot.py]: Delete this commented out line.

File: src/herbpy/tsr/block.py [hunk 8]
```diff
@@ -1,7 +1,47 @@
...
+    alpha = 0.8 # orientation of end-effector relative to block
+
+    block_in_world = block.GetTransform()
+    block_in_world[:3,:3] = numpy.eye(3) # ignore orientation
+    ee_in_block = numpy.array([[numpy.cos(alpha), 0., -numpy.sin(alpha), 0.3*numpy.sin(alpha)+0.04],
+                               [ 0., -1, 0, 0.],
+                               [-numpy.sin(alpha), 0., -numpy.cos(alpha), 0.25+offset],
+                               [ 0., 0., 0., 1.]])
+    Bw = numpy.zeros((6,2))
+    Bw[5,:] = [-numpy.pi, numpy.pi-.0001]
```
Comment (mkoval) [inline src/herbpy/tsr/block.py]: This `- 0.001` should not be required as of https://github.com/personalrobotics/prpy/pull/194.

File: src/herbpy/tsr/block.py [hunk 9]
```diff
@@ -43,4 +83,41 @@ def block_at_pose(robot, block, position, manip=None):
...
+
+    block_pose = block.GetTransform()
+    block_pose[:3,:3] = numpy.eye(3) # ignore orientation
+    ee_in_block = numpy.dot(numpy.linalg.inv(block_pose), manip.GetEndEffectorTransform())
+    Bw = numpy.zeros((6,2)) 
+    Bw[2,:] = [0., 0.04]  # Allow some vertical movement
+   
+    for tsr in pose_tsr_chain.TSRs:
+        if tsr.manipindex != manip_idx:
+            raise Exception('pose_tsr_chain defined for a different manipulator.')
```
Comment (mkoval) [inline src/herbpy/tsr/block.py]: This should be a `ValueError` since it's an error in one of the arguments.

File: src/herbpy/tsr/block_bin.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    Tw_e[2,3] = 0.17 # set the object on top of the bin - bin is 13cm high
+
+    Bw = numpy.zeros((6,2))
+
+    xdim = max(0.085 - padding, 0.0)
+    ydim = max(0.135 - padding, 0.0)
+    Bw[0,:] = [-xdim, xdim ] # move along x and y directions to get any point on tray
+    Bw[1,:] = [-ydim, ydim]
+    Bw[2,:] = [-0.02, 0.04] # verticle movement
+    Bw[5,:] = [-numpy.pi, numpy.pi-.0001] # allow any rotation around z - which is the axis normal to the tray top
```
Comment (mkoval) [inline src/herbpy/tsr/block_bin.py]: See above. This `- 0.0001` should no longer be necessary.

**Comment by mkoval** [comment]
This looks pretty good. I'm good to merge this as soon as the few nitpicks above are addressed. :smile:

File: src/herbpy/action/blocks.py (hunk 2 above)
Comment (jeking04) [inline src/herbpy/action/blocks.py]: Addressed.

File: src/herbpy/action/blocks.py (hunk 3 above)
Comment (jeking04) [inline src/herbpy/action/blocks.py]: Addressed.

File: src/herbpy/action/blocks.py (hunk 4 above)
Comment (jeking04) [inline src/herbpy/action/blocks.py]: I don't understand this one or the next. Where should I be locking?

File: src/herbpy/action/blocks.py (hunk 5 above)
Comment (jeking04) [inline src/herbpy/action/blocks.py]: Done.

File: src/herbpy/action/blocks.py (hunk 4 above)
Comment (mkoval) [inline src/herbpy/action/blocks.py]: I thought we decided that the functions in `tsrlibrary` can safely assume that the environment's locked. If that is the case, we need to lock the environment whenever calling `tsrlibrary`.

File: src/herbpy/tsr/block_bin.py (hunk 10 above)
Comment (jeking04) [inline src/herbpy/tsr/block_bin.py]: Verified it is no longer necessary and removed.

File: src/herbpy/tsr/block.py (hunk 8 above)
Comment (jeking04) [inline src/herbpy/tsr/block.py]: Verified it is no longer necessary and removed.

File: src/herbpy/herbrobot.py (hunk 7 above)
Comment (jeking04) [inline src/herbpy/herbrobot.py]: I removed the comment. @ClintLiddick was looking into how to fix this properly.

File: src/herbpy/action/blocks.py (hunk 1 above)
Comment (jeking04) [inline src/herbpy/action/blocks.py]: Addressed.

File: src/herbpy/tsr/block.py (hunk 9 above)
Comment (jeking04) [inline src/herbpy/tsr/block.py]: Addressed.

**Comment by jeking04** [comment]
@mkoval Okay I think I got everything except for the two lock comments I didn't understand.

File: src/herbpy/action/blocks.py (hunk 4 above)
Comment (jeking04) [inline src/herbpy/action/blocks.py]: Hmm...I thought we actually decided the opposite. We have been adding locks to the tsrlibrary wherever we find missing ones. I am not opposed to going with this paradym instead, but I think we should do it in a separate issue/PR where we carefully go through the library and remove or cleanup locking.

Comment (mkoval) [inline src/herbpy/action/blocks.py]: That's fine by me.

@jeking04 Can you create an issue for that?

**Pull request merged**